### DISCUSSION TITLE
A sketch with multiple windows?

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#158] [Q&A] [answered]
I have this minimal example for Processing Python mode: 

```python
def setup():
    size(200, 300)
    second_window = OtherWindow("2nd")  
    
def draw():
    background(0)
    ellipse(mouseX, mouseY, 10, 10)


class OtherWindow(PApplet):  
        
    def __init__(self, title=""):
        switches = ('--sketch-path=' + sketchPath(), '')
        PApplet.runSketch(switches, self)  
        self.surface.setTitle(title)
        
    def settings(self):
        self.size(300, 200)
        
    def draw(self):  # draw for the second window
        self.background(255)
        self.fill(0)
        self.rect(self.mouseX, self.mouseY, 10, 10)
```

Will this be possible in py5 with *imported mode* / *module mode*, or just with *class mode*?
Let's talk about it here...

This is my first attempt at *class mode*, no luck yet, I guess I'll need some help...

```python
from py5 import Sketch

class SketchySketch(Sketch):
    def __init__(self, title=""):
        if title:
            self.window_title = title
        self.title = title
        print(f"I'm {self.title}'s __init__!")
        super().__init__()

    def settings(self):
        self.size(300, 200)
        print(f"I'm {self.title}'s settings!")

    def draw(self):  # este é o draw pra a segunda janela
        if self.title == 'B':
            self.background(255)
            self.fill(0)
        else:
            self.background(0)
            self.fill(255)
        self.rect(self.mouse_x, self.mouse_y, 10, 10)

a = SketchySketch('A')
b = SketchySketch('B')
a.run_sketch()
b.run_sketch()  # I also tried this inside the __init__ mimicking the Py.Processing example 
```
Looks like the second sketch is initialized but does not run.

![image](https://user-images.githubusercontent.com/3694604/188509122-81cd40b0-3515-4724-a4b6-02f45b77c9e5.png)

UPDATE: The second sketch `b` runs after `a` is closed!

**Comment by villares** [comment]
Some progress, I suppose... `run_sketch(block=False)` seems to make it work....

```python
from py5 import Sketch

class SketchySketch(Sketch):
    def __init__(self, title=""):
        self.window_title = title
        print(f"I'm {self.window_title}'s __init__!")
        super().__init__()
        self.run_sketch(block=False)            

    def settings(self):
        self.size(300, 200)
        print(f"I'm {self.window_title}'s settings!")

    def setup(self):
        if self.window_title == 'B':
            self.c = SketchySketch('C')
        print(f"I'm {self.window_title}'s setup!")


    def draw(self):  # este é o draw pra a segunda janela
        if self.window_title == 'B':
            self.background(255)
            self.fill(0)
        elif self.window_title == 'C':
            self.background(0, 0, 100)
            self.fill(0, 200, 0)
        else:
            self.background(0)
            self.fill(255)
        self.rect(self.mouse_x, self.mouse_y, 10, 10)

a = SketchySketch('A')
b = SketchySketch('B')
```
And I could get `b` to start `c` and keep a reference to it, so maybe I can make them act in coordination.

![image](https://user-images.githubusercontent.com/3694604/188511331-b443af83-30ea-4c62-8e99-d5a35dd9eb2a.png)

**Comment by hx2A** [comment]
Interesting that this is valid Processing Python code:

```python
class OtherWindow(PApplet):  
        
    def __init__(self, title=""):
        switches = ('--sketch-path=' + sketchPath(), '')
        PApplet.runSketch(switches, self)  
        self.surface.setTitle(title)
```

You can call `runSketch()` from within `__init__()`? That wouldn't work in py5. You also can't set the surface title there because the surface doesn't exist until after `settings()` has been executed.

Also, the variable `window_title` conflicts with the `Sketch.window_title()` method.

In your `setup()`, why are you creating an instance of `SketchySketch`? I don't think it is a good idea to instantiate new Sketch instances from within another Sketch.

Try this instead:

```python
from py5 import Sketch

class SketchySketch(Sketch):
    def __init__(self, title=""):
        super().__init__()
        self._window_title = title
        print(f"I'm {self._window_title}'s __init__!")

    def settings(self):
        self.size(300, 200)
        print(f"I'm {self._window_title}'s settings!")

    def setup(self):
        self.window_title(self._window_title)
        print(f"I'm {self._window_title}'s setup!")

    def draw(self):  # este é o draw pra a segunda janela
        if self._window_title == 'B':
            self.background(255)
            self.fill(0)
        elif self._window_title == 'C':
            self.background(0, 0, 100)
            self.fill(0, 200, 0)
        else:
            self.background(0)
            self.fill(255)
        self.rect(self.mouse_x, self.mouse_y, 10, 10)

a = SketchySketch('A')
b = SketchySketch('B')
c = SketchySketch('C')

a.run_sketch()
b.run_sketch()
c.run_sketch()
```

This should work better.

**Reply by villares** [reply]
Hmm, how strange.... My second attempt worked fine as shown. It did work to call `sketch_run()` inside `__init__()`.  What made the difference for me was `block=False`.

Your code without `block=False` doesn't work for me: the next window will only appear after the previous one is closed.

I see how it could get messy to create one from the other, I wanted to see if it worked, and it did. But I guess, in the back of my mind,  I'd like to communicate the idea of a single sketch that controls a second one like the original example. I guess that the code "outside" that instantiates the sketches can deal with making one of them have a reference to the other & etc.

**Reply by hx2A** [reply]
Are you running this with a generic Python interpreter (or with Thonny)? When you run it that way, `run_sketch()` defaults to `block=True`. That means the next line of code won't run until after the Sketch exits. When  you run it through IPython or Jupyter Notebook, the default is `block=False`. You need `block=False` for it to run Sketches b and c.

> It did work to call sketch_run() inside __init__()

I can see how it would work, but make sure it is the last line in the `__init__()` method.

**Reply by villares** [reply]
> Are you running this with a generic Python interpreter (or with Thonny)?

I was running with *imported mode* off, which should be just like running from the terminal (which I tested now), maybe we should double-check the block default? 

![image](https://user-images.githubusercontent.com/3694604/188517139-c5c3beaf-abde-4152-9a8f-7c92cc456a90.png)

**Reply by villares** [reply]
and if I remove the `block=False`,  I get the blocking behavior...

**Reply by hx2A** [reply]
> maybe we should double-check the block default?

I can see how this can be confusing. I thought that if someone is running it as a script, such as

```bash
$ python my_code.py
```

if the user forgets to include `block=True`, the Python interpreter will exit and kill the Sketch. But in Jupyter notebook, `block=True` is rarely what you want, because you can keep running other notebook cells while the Sketch is running. It's also never what you want on OSX + Jupyter, where blocking isn't even allowed (because the Sketch will not work).

OSX complicates things. On OSX in Jupyter, `block=False` is the only option and only way py5 can work. On OSX outside of Jupyter, in the Python interpreter, `block=True` is the only option and the only way py5 can work.

I can see how we might want to make one or the other always the default, regardless of how it is run, but doing so isn't possible because of OSX. The way it works now is at least closer to being consistent across platforms, and that's important also.

**Reply by villares** [reply]
So, just to confirm, `run_sketch()`'s default for module mode & class mode on a standalone script (not a Jupyter notebook) is `block=True`. No problem.

**Reply by hx2A** [reply]
> So, just to confirm, `run_sketch()`'s default for module mode & class mode on a standalone script (not a Jupyter notebook) is `block=True`. No problem.

That's right. Perhaps a more concise way to explain it is when run through Jupyter, `block=False` is the default and is the only option for OSX. When not run through Jupyter, `block=True` is the default and the only option for OSX.

**Comment by villares** [comment, marked as answer]
So this is my silly example, for now...

```python
from py5 import Sketch

class SketchySketch(Sketch):
    def __init__(self, title="", other=None):
        self.other = other
        self.title = title
        self.clicked = False
        super().__init__()        

    def settings(self):
        self.size(300, 200)

    def setup(self):
        self.window_resizable(True)
        if self.title:
            self.window_title(self.title)
        if self.other:
            self.other.other = self

    def draw(self):
        if self.title == 'A':
            self.background(255)
            self.fill(0)
        else:
            self.background(0)
            self.fill(255)
        self.rect(self.mouse_x, self.mouse_y, 10, 10)
        
        if self.clicked:
            w, h = self.width, self.height
            self.fill(128, 128)
            self.circle(w / 2, h / 2, min(w, h) * 0.8)
            
    def mouse_pressed(self):
        self.clicked = not self.clicked
        if self.other:
            self.other.clicked = self.clicked

a = SketchySketch('A')
a.run_sketch(block=False)
b = SketchySketch('B', a)
b.run_sketch(block=False)
```

![image](https://user-images.githubusercontent.com/3694604/188522937-b9835d7e-10ea-4c25-9ff5-7a916533b537.png)

**Reply by hx2A** [reply]
Interesting. Do you want the Sketches to be able to communicate with each other? Clicking one is like clicking both? What you have there should work fine.

**Reply by villares** [reply]
Yes, that's it. The original Py.Processing example was in that direction, something like "a sketch with two windows". It can be good for keeping a GUI in a screen and the drawing in another, for instance.

**Reply by hx2A** [reply]
Interesting.

If this code is something you will share, you should mention in the comments that OSX users should run it through Jupyter.

**Reply by villares** [reply]
Oh, I see, in MacOS it will break. I guess I could spawn a thread in MacOS?

**Reply by hx2A** [reply]
> Oh, I see, in MacOS it will break. I guess I could spawn a thread in MacOS?

I've tried everything! Spawning a thread won't work. You could spawn a separate process though, and use the tools in the multiprocessing library for the two Sketches to communicate with each other. That is a bit of a drag to use though.

**Reply by villares** [reply]
Ok, I'll add a warning. And I'll try again contacting the Swift Processing guy to see if he has any clues.

**Reply by hx2A** [reply]
>  i'll try again contacting the Swift Processing guy to see if he has any clues.

Who is the Swift Processing guy?

**Reply by villares** [reply]
@masoodkamandy! I hope I was not being rude calling him the Swift Processing guy!
I meant the very nice guy I met on Twitter who is involved with Swift Processing.

**Reply by bluekamandy** [reply]
I wish I could help, but this is sadly outside my area of expertise. Ask me about Xcode or Swift and I'm your guy. And I don't mind being called the SwiftProcessing guy! 😂

**Reply by hx2A** [reply]
@masoodkamandy, thanks for stopping by. You created a version of Processing for Swift? It's nice to meet you. I created a version of Processing for Python. It's so rewarding to have created such a thing, as you must know!

**Reply by bluekamandy** [reply]
It's changed my life, actually, so I totally understand! Understanding all of the decisions behind Processing and exploring why or why they do not fit in a new language has been one of the highlights of my creative life. Nice to meet you!

**Reply by hx2A** [reply]
> It's changed my life, actually

I'd like to hear more about that!

I'm certainly learning a lot from others in the py5 community about teaching coding to beginners. It's interesting and eye opening.
